### PR TITLE
tcp + rxm: optimization path for small tagged messages

### PR DESCRIPTION
**Pull request by shefty**
Work in progress:

This series adds a send-side tagged API implementation to tcp.  This is an unpublished optimization path, as tcp does not actually support tagged messages, particularly, on the receive side.  TCP does not implement a receive side tagged queue.

The use is that rxm can call the tcp tagged api in order to send small (eager sized) tagged messages.  By doing so, tcp carries the tag in its header, eliminating the rxm header completely.  For very small messages, this cuts the network overhead by more than 50%.  TCP will report the tag with the CQ completion.

This feature can only be enabled if dynamic receive buffers are enabled.  When tcp invokes the callback to rxm at the receiving side, rxm can use the completion data to construct what would have been the rxm header.  The construction of the rxm header isn't technically necessary, but does minimize the code changes that are needed.  

The tcp changes are complete.  The receive side rxm changes are in place.  Still to do are the rxm send changes, so that it calls the tcp tsend calls.

### CODE REVIEW

File: prov/tcp/src/tcpx_ep.c [hunk 1]
```diff
@@ -61,6 +63,11 @@ void tcpx_hdr_bswap(struct tcpx_base_hdr *hdr)
 		ptr += sizeof(uint64_t);
 	}
 
+	if (hdr->flags & TCPX_TAGGED) {
+		*((uint64_t *)ptr) = ntohll(*((uint64_t *) ptr));
```
Comment (ooststep) [inline prov/tcp/src/tcpx_ep.c:67]: can TCPX_REMOTE_CQ_DATA and TCPX_TAGGED both be set?  If so, could this be reversing the above ntohll swap?  if not, should this be an `else if` to make it obvious that both can't be true?

Comment (shefty) [reply prov/tcp/src/tcpx_ep.c:67]: Both flags can be set.  ptr is incremented after swapping the cq data.  Same after swapping the tag.  It's the current location in the header being processed.

**Comment by shefty** [comment]
Latest update added rxm send support.  Inject path is still not supported.  Testing has not been confirmed yet.

File: prov/tcp/src/tcpx_ep.c (hunk 1 above)
Comment (ooststep) [reply prov/tcp/src/tcpx_ep.c:67]: got it. thanks!

**Comment by shefty** [comment]
Added inject path support.  Also a few stray updates to tcp inject size and buffer pools.  Original set of patches pushed to PR should be mostly unmodified.

**Comment by shefty** [comment]
@ooststep - This series is getting long, and I have another 10 or so changes identified beyond these that I still need to write.  I've done some simple testing to verify the new functionality works, but it is still disabled by default.  The overall risk here should be minimized.  Rather than pile onto this PR, I would prefer to merge these.  It looks like you reviewed some of the changes already.  I'd appreciate it if you could review the rest prior to me merging.

**Pull request merged**